Bug fix slow down smoothly - #50

alrapal opened the pull request:
##########################################################
**Type**:
bug fix

**Tags**:
resolves #50
closes #50

**Changes**:
- method has been changed so allows user input. It also keeps other reactions possible (emergency break and side avoidance)
- Reorganised the code so the helpers are coming after the methods they are helping.
- All methods have a comment block explaining their use
- Reimplementation of handleInput to ease testing. Instruction given above the setup()

**Notes**:
- coded with @iachella & @gusaxtcha

alrapal committed:
Bug fix slow down smoothly - #50
##########################################################
Type:
bug fix
Tags:
#50
Changes:
- method has been changed so allows user input. It also keeps other reactions possible (emergency break and side avoidance)
Notes:
- coded with @iachella & @gusaxtcha

ediz-genc reviewed (changes requested):
I have tried the code with several scenarios/cases and created the situation as it is stated in issue #50 .

It seems like there is no consistency according to issue #50. It succeeds in stopping only 2 times while smoothly slowing down if another car jumps in front. I didn't come across any side avoidance with this code snippet. 

I assume you might wanna test the cases and modify them further in terms of the right solution to issue #50. 👍🏻

ediz-genc commented on arduino/smartcar/smartcar.ino:
```diff
@@ -242,6 +308,7 @@ void reactToSides() {
         }
     }
 }
+
```
in the sideavoidance(), there are possible data loss from int getDistance() to float.

ediz-genc commented on arduino/smartcar/smartcar.ino:
```diff
@@ -220,11 +254,43 @@ bool reactToSensor(int sensorDistance, int STOP_DISTANCE){
 /**
  * Method to look and compare two sensor values (100ms apart) to know if an obstacle is coming towards the car
  */
 void reactToSides() {
```
in the reactToSides(), there are possible data loss from int getDistance() to float.

ediz-genc commented on arduino/smartcar/smartcar.ino:
```diff
@@ -158,9 +155,44 @@ void handleAngleTopic(int input){
- * handleInput helper method for speed
+ * helper for MQTT and Serial inputs taking care of the speed
  */
 void handleSpeedInput(float distance, int inputSpeed){
```
in handleSpeedInput() if-statement, comparing int with float. possible data loss might cause/trigger some other inconstancy in reaction.

ediz-genc reviewed (changes requested):
There are some possible data losses between int methods (such as getDistance()) to float. That might cause/trigger inconstancy or not :) It might need some other way of thinking in methods :)

alrapal commented:
> I have tried the code with several scenarios/cases and created the situation as it is stated in issue #50 .
> 
> It seems like there is no consistency according to issue #50. It succeeds in stopping only 2 times while smoothly slowing down if another car jumps in front. I didn't come across any side avoidance with this code snippet.
> 
> I assume you might wanna test the cases and modify them further in terms of the right solution to issue #50. 👍🏻

Regarding the side avoidance, do you mean you did not try it or did not succeed in making it work? Using a car to make the side avoidance is very tricky because of the nature of the sensors and the design of the car (the sensors are detecting just the upper square from the car which implies to have the car very close to trigger the avoidance. Emulator limitation. If you want to try the side avoidance, try to use a wall instead. The ones being behind the car when you start the script is the best.)

When trying side avoidance, it works on my computer. 

The emergency break seems to suffer from the range being too short. I will work on fix.
I will also change the floats to be int. There should not be a loss of data since floats are more precise than int but it makes the code cleaner for sure. 

Thanks again for your very complete feedback =)

alrapal committed:
review fix slow down smoothly - #50
##########################################################
Type:
bug fix
Tags:
#50
Changes:
- Increased the range to allow the car to react to car coming at the front
- changes values to int to avoid loss of data
- removed unused constants
- added a check with the frontIR in emergency break to increase chances of reaction
Notes:
- changes based on @edizgenc2021 review for this PR

ediz-genc commented:
I tried with the case car approaching from side. It didnt succeed the side avoidance. I will have a look for the wall case later on today.

I agree with the limitation of emulator. :) The situations/cases cant be created as many as we would like to have, but car approaching from side case is pretty accurate case for such a sensor. I mean side avoidance should function if another car approaching from side. You may keep in mind to test it with this case as well 🙂

alrapal commented:
> I tried with the case car approaching from side. It didnt succeed the side avoidance. I will have a look for the wall case later on today.

You tried it like the figure below. But this is not a realistic case. If a car is coming at high speed on your side, you cannot avoid that. Even if you start an avoidance, the car will hit you (just because of the way a car drives.
![Capture d’écran 2021-04-24 à 13 08 35](https://user-images.githubusercontent.com/71485862/115956850-7f0b6100-a4ff-11eb-877d-9bda2612a981.png)

The idea behind the side avoidance is if you have something, coming towards you at a avoidable angle like below. 
![Capture d’écran 2021-04-24 à 13 11 12](https://user-images.githubusercontent.com/71485862/115956906-c691ed00-a4ff-11eb-86e8-012a5c09bb9a.png)


> I agree with the limitation of emulator. :) The situations/cases cant be created as many as we would like to have, but car approaching from side case is pretty accurate case for such a sensor. I mean side avoidance should function if another car approaching from side. You may keep in mind to test it with this case as well 🙂

The limitation I am talking about are the following: 
In the code we are using IR sensors that have a range of 25cm to 120cm. But in the emulator, it only detects at a range of 40 since the first value you get from the sensors are 40. like shown in both screenshots below. 
Note also on the first screenshot that the sensor are detecting the top of the car, and not the front, which contributes to the limitation I was mentioning. 
![Capture d’écran 2021-04-24 à 13 05 34](https://user-images.githubusercontent.com/71485862/115957055-a0b91800-a500-11eb-8791-fc6c5ca2f58e.png)
![Capture d’écran 2021-04-24 à 12 59 29](https://user-images.githubusercontent.com/71485862/115956965-0f49a600-a500-11eb-8e82-b709897e202b.png)

Because of all this, you cannot expect the car to avoid from the side when it comes to the way you tested it (90° angle). I suggest you try as I described before:
- Coming towards a wall at a certain angle 
- or using the controllable car to derive towards our car 

I hope this is clearer @edizgenc2021  =)

ediz-genc commented:
Oh, man :joy: :joy: it is getting hot here. Joke aside. It was clear all along. What I showed you in the records earlier was not definitely replicating [T-bone side collision with high speed](https://www.fordlawoffices.com/side-impact-car-accidents.html), on the contrary, to test the side sensors with coverage of [GP2Y0A02](https://platisd.github.io/smartcar_shield/_g_p2_y0_a02_8cpp_source.html) (IR 25-120). 

Recall the records I have sent you about side avoidance -> against moving object (i.e. a car which tries to overtake, even your second picture situation); 

https://user-images.githubusercontent.com/80987284/115969643-3d4ddb00-a53e-11eb-8e7a-328a135a2fb5.mp4

[Estimated side car accidents 25%](https://www.fordlawoffices.com/side-impact-car-accidents.html) it would be pretty accurate and realistic if you take it as a case.

Recall also the issue of yours as you are the author; 

issue #50 
**"Current behavior"**
When the car starts to slow down smoothly, it does not detect **obstacles coming from the side** or popping at the front

**Expected behavior**
The car should priorities breaking to slowing down smoothly **and continue reacting to obstacles coming from the side.**"

A Wall doesn't move or comes from the side of the car. Maybe you need to modify your wording in issue #50 .

In the end, I should do as we are told by @platisd a month ago, reviewing is not just approving the PR. We need to take into account the most important cases in functions. We are all here to learn from each part of the courses ;)

alrapal commented:
No not hot, don't worry!
I know you did not come at high speed from the side, I meant more that in the case of a car coming at the side, there is no way to avoid that (at high speed or like the recording you sent) because a car cannot jump 😂 (any other move would not avoid the accident: turning, stopping or accelerating, especially with such short sensors).

I was also pointing out a bug in the emulator that does not take the theoretical range for the sensors and limit them at 40, added to half of the car not being detected, which makes it even more impossible to detect / react on time to try any manoeuvre if the comes at a 90degree angle.

[EDIT] : This is not a bug as @RuthgerD  pointed out below and is documented on their wiki. But it is still limiting the detection of another coming car. 

I guess the wording can be confusing and imply that the obstacle is moving (which is partly covered with the case of a car deriving towards the véhicule). 

How would you suggest changing the wording? 

Should we change for something like "when an obstacle is detected on time at the sides or at the front to break in emergency or try to avoid the obstacle" ? 

Lastly, I appreciate you take time to do the review and never implied in any way you should accept this PR without asking questions. I also appreciate you changed your original message and apologise if mine could be persevere as "hot", I was just trying to explain what I meant with limitations and why the way you tested the feature could not work even if I wanted it to work while keeping the comment as straight to the point as possible. 

Anyway, we can talk about this on Monday! Have a nice week end 😊

ediz-genc commented:
Yeah, ok. No worries. You do as you like with your issue or your functions. Just make sure that issues you have and what you really like to achieve with your functions serve the same purpose. 
Happy nice weekend as well.

ediz-genc commented:
On the other hand, I/others can help you on this. I mean we all can decide together on Mondays meeting👍🏻

ChrisAxt commented:
@edizgenc2021, Good morning! :) Regarding the last video you send, We could definitely try handle a case like that. There are a few things to consider though:
1) The other car is only detected at about 8 seconds into the video (This is estimated from what I can see), at that point the other car is extremely close and in reality I can't imagine what could be done to avoid that.
2) What can our car do to avoid that, especially since the other car is moving. Maybe it's more like a brace for impact kind of situation?
3) "Smart"/ autonomous cars can't avoid every accident unless all cars are "Smart"/ autonomous and even then there might be situations that the code "chooses" the lesser of two accidents.

If you have any suggestions regarding the above, I would love to hear and maybe brainstorm some ideas too :)
I understand your concerns about the way the bug description was worded and we will work on that too.

ediz-genc commented:
@gusaxtcha, Hej Chris. Well, the initial video has been recorded on the basis of the description for issue #50 (against a moving object) with two related main reasons;

1. The second car gets closer and closer from side to get into the sensor coverage of the 1st car to determine if it is going to react by starting to turn in the same direction (for an example) as the 2nd car (basically to see if the functions are accurately created based on the issue description). This would help to see if the side sensors are functioning. If it reacts to a moving object it would react the same way for stationary objects.
2. In case, if this was the case as in issue #50 description then the sensor should preferentially act to another case if the car gets closer to another car that has a "side avoidance sensor", (i.e a car that overtakes another car with a side avoidance sensor) then it reacts as above and starts to turn the opposite direction as possible by slowly moving. 

__To your considerations_ (because I like brainstorming :) );_
1. Well, together with the above explanations, in reality, it could have been avoided (partially (not definitely 100%) to minimize the impact) against a moving object, let's say overtaking 2nd vehicle, (i) reinforced airbags inside the door panel in several places opens outwards with specific diameter to avoid/"minimize" the impact, (ii) if there is enough space (by detecting from/back sensors) at the front or back of the vehicle with side avoidance sensor would speed up/break to avoid the impact. More cases or avoidance can be created in real life.
2. I am kind of agree with you. It would have turned in the direction of other vehicles that come from the side to avoid/"minimize" the inevitable impact or speed up/break suddenly to escape from T-bone side collision.
3.  If all cars are autonomous or smart, then probably accidents would never happen and we wouldn't discuss here above case :) or then the accident would happen only if we have t.ex. MQTT connection malfunction :) between sensors and accident avoidance systems.

On the other hand, I believe that this is not the right channel for brainstorming (otherwise this issue will stay open until the end of the project), we can take this brainstorming to another communication channel, verbally(but shortly) in our meeting maybe. Regarding the emulator, we all know that wording of the issue and functions was not on the same path. In the emulator, we can/could cover fewer cases than in the real world. The most important consideration we should take into account is to get the issues which would help for the purpose of our project.

ChrisAxt commented:
@edizgenc2021 Great! I agree, we can chat more about this in our meeting, If we need more time we can always schedule another meeting specifically for this issue. :)

I will respond to this message properly when we talk. Glad you like brainstorming too :D

RuthgerD commented:
> a bug in the emulator

If you believe there is a bug in the emulator I invite you to report such things.

According to the [wiki](https://github.com/ItJustWorksTM/smce-gd/wiki/Vehicle-Capabilities#infrared-distance-sensor) we define all 3 sensors to be the same capping out at 40 cm which is already more than you would get in the real world as the real car uses the `GP2D120` IR which caps at [25 cm](https://github.com/platisd/smartcar_shield/blob/master/src/sensors/distance/infrared/analog/sharp/GP2D120.cpp#L6).

If you think IR capabilities should change discuss with Dimitris.

alrapal commented:
> > a bug in the emulator
> 
> If you believe there is a bug in the emulator I invite you to report such things.
> 
> According to the [wiki](https://github.com/ItJustWorksTM/smce-gd/wiki/Vehicle-Capabilities#infrared-distance-sensor) we define all 3 sensors to be the same capping out at 40 cm which is already more than you would get in the real world as the real car uses the `GP2D120` IR which caps at [25 cm](https://github.com/platisd/smartcar_shield/blob/master/src/sensors/distance/infrared/analog/sharp/GP2D120.cpp#L6).
> 
> If you think IR capabilities should change discuss with Dimitris.

Hej @RuthgerD, 
Thanks for stopping by. It is effectively mentioned in the wiki that the limit is 0.4 but I find it confusing to implement the 3 sensors since their interest is their different range. Is it so that the 25 to 120 is a 25 to 40 cm in the emulator ? or is the lower range at 0 for all? (Actually that is also documented on the wiki and there is no minimal range). So there is no differences between any IR sensor in the emulator =). Need to read more carefully. 

[EDIT: corrected wrong value - added answer to question]

Also, would it be possible to lower the sensors so they detect the car sooner? As you may have seen on the screenshots, the IR sensors is detecting the top of a car only. 
Would you rather like that we create an issue on your git to suggest that change ?
Thanks for your response =)

RuthgerD commented:
> confusing to implement the 3 sensors

Yes I agree, it was done since the sketch side implementation came for free for all 3 as the emulator controls all the limits.

>  Is it so that the 25 to 120 is a 25 to 40 cm in the emulator ?

Not sure what you mean with this, but at the moment of speaking the range is 0 - 0.4m for all IR sensor types. 
its set at 0.4m because 0.25m is truly quite small and I foresaw people complain about it :sweat_smile: 

> would it be possible to lower the sensors so they detect the car sooner?

I think its better to angle the sensor down (the real car attaches the sensors on top so lowering it any more would be unrealistic)

Dimitris actually asked for a way to configure the sensors so maybe angle could be a property, haven't started on that yet though.

alrapal commented:
> > confusing to implement the 3 sensors
> 
> Yes I agree, it was done since the sketch side implementation came for free for all 3 as the emulator controls all the limits.

Ok I see! Thanks for that clarification. 
> > Is it so that the 25 to 120 is a 25 to 40 cm in the emulator ?
> 
> Not sure what you mean with this, but at the moment of speaking the range is 0 - 0.4m for all IR sensor types.
> its set at 0.4m because 0.25m is truly quite small and I foresaw people complain about it 😅

I got the answer by looking at the next sentence in your documentation 😂 need to read more carefully 

> > would it be possible to lower the sensors so they detect the car sooner?
> 
> I think its better to angle the sensor down (the real car attaches the sensors on top so lowering it any more would be unrealistic)

That would be a great solution yes ! Hope we get to see it ! But no pressure haha

> Dimitris actually asked for a way to configure the sensors so maybe angle could be a property, haven't started on that yet though.

alrapal commented:
#50 is updated so it is (hopefully) not confusing anymore :D

iachella reviewed (approved):
This pull request has been discussed by the team, and there have been some clarifications:

**As clarification about the bug that this PR is fixing:**
- The issue has been updated (see #50 Expected behaviour for info)

**As clarification about the possible data loss:**
- The getDistance() methods for IR sensors return int values, therefore they can be stored in int variables
- The compilers might still complain about this

Good work with reorganizing the code and the comments :)

alrapal commented:
> * The compilers might still complain about this

I checked and the compiler is not complaining about the int to float conversion ! So it should be ok 😄

ediz-genc commented:
![image](https://user-images.githubusercontent.com/80987284/116393184-99a44980-a821-11eb-9c28-8a3aff44a3b8.png)


After the meeting has been held today by me(@edizgenc2021) _"reviewer"_, Chris (@gusaxtcha) _"contributor"_, and Alex (@alrapal) _"contributor"_, modification of wording/re-phrasing of issue #50 as above screenshot and as well as the code modification which has been done, the bug seems to be fixed and consistency between issue and implemented feature holds.

Good job everyone 👍🏻  :)
